wasm/jit/engine.go

```diff
@@ -15,70 +15,327 @@ import (
 	"github.com/tetratelabs/wazero/wasm/internal/wazeroir"
 )
 
-type engine struct {
-	// These fields are used and manipulated by JITed code.
-
-	// The actual Go-allocated stack.
-	// Note that we NEVER edit len or cap in JITed code so we won't get screwed when GC comes in.
-	stack []uint64
-	// Stack pointer on .stack field which is accessed by [stackBasePointer] + [stackPointer].
-	stackPointer uint64
-	// stackBasePointer is set whenever we make function calls.
-	// Background: Functions might be compiled as if they use the stack from the bottom,
-	// however in reality, they have to use it from the middle of the stack depending on
-	// when these function calls are made. So instead of accessing stack via stackPointer alone,
-	// functions are compiled so they access the stack via [stackBasePointer](fixed for entire function) + [stackPointer].
-	// More precisely, stackBasePointer is set to [callee's stack pointer] + [callee's stack base pointer] - [caller's params].
-	// This way, compiled functions can be independent of the timing of functions calls made against them.
-	stackBasePointer uint64
-	// Where we store the status code of JIT execution.
-	jitCallStatusCode jitCallStatusCode
-	// Set when statusCode == jitStatusCall{Function,BuiltInFunction}
-	// Indicating the function call index.
-	functionCallAddress wasm.FunctionAddress
-	// Set when statusCode == jitStatusCall{Function,BuiltInFunction}
-	// We use this value to continue the current function
-	// after calling the target function exits.
-	// Instructions after [base+continuationAddressOffset] must start with
-	// restoring reserved registeres.
-	continuationAddressOffset uintptr
-	// The current compiledFunction.globalSliceAddress
-	globalSliceAddress uintptr
-	// memorySliceLen stores the address of the first byte in the memory slice used by the currently executed function.
-	memorySliceAddress uintptr
-	// memorySliceLen stores the length of memory slice used by the currently executed function.
-	memorySliceLen uint64
-	// The current compiledFunction.tableSliceAddress
-	tableSliceAddress uintptr
-	// tableSliceLen stores the length of the unique table used by the currently executed function.
-	tableSliceLen uint64
-	// Function call frames in linked list
-	callFrameStack *callFrame
-	// callFrameNum tracks the current number of call frames.
-	// Note: this is not len(callFrameStack) because the stack is implemented as a linked list.
-	callFrameNum uint64
-
-	// The following fields are only used during compilation.
-
-	// Store the compiled functions.
-	compiledFunctions map[wasm.FunctionAddress]*compiledFunction
-}
+type (
+	engine struct {
+		// There contexts are read and written by JITed code.
+		// Note that we embed these structs so we can reduce the costs to access fields inside of them.
+		// Also, that eases the calculation of offsets to each field.
```

Comment:
to ensure this isn't conflated with the godoc on globalContext
```suggestion
		// Also, that eases the calculation of offsets to each field.

```